fix(compiler): improved support for `:host-context()` selectors

petebacondarwin opened the pull request:
Two commits that fix problems with emulated shadow dom handling of `:host-context`.

Fixes #19199
Fixes #14349

petebacondarwin committed:
fix(compiler): support multiple `:host-context()` selectors

In `ViewEncapsulation.Emulated` mode, the compiler must generate additional
combinations of selectors to handle the `:host-context()` pseudo-class function.

Previously, when there is was more than one `:host-context()` selector in a
rule, the compiler was generating invalid selectors.

This commit generates all possible combinations of selectors needed to
match the same elements as the native `:host-context()` selector.

Fixes #19199

petebacondarwin commented on packages/compiler/src/shadow_css.ts:
```diff
@@ -534,11 +539,12 @@ const _cssContentUnscopedRuleRe =
-const _cssColonHostContextRe = new RegExp('(' + _polyfillHostContext + _parenSuffix, 'gim');
+const _cssColonHostRe = new RegExp(_polyfillHost + _parenSuffix, 'gim');
+const _cssColonHostContextReGlobal = new RegExp(_polyfillHostContext + _parenSuffix, 'gim');
+const _cssColonHostContextRe = new RegExp(_cssColonHostContextReGlobal, 'im');
```
Note later on we cannot use the `_cssColonHostContextReGlobal` because the `global` flag makes the regex stateful, so it remembers its last position.

JoostK commented on packages/compiler/src/shadow_css.ts:
```diff
@@ -650,3 +656,45 @@ function escapeBlocks(
+      combined[i] = contextSelector + combined[i];
+    }
+  }
+  return combined.filter(s => s !== undefined)
```
How could `s` be undefined? `combined` is declared as `string[]` and the values that are written to it appear to always be strings, so I suppose this filter is not needed?

petebacondarwin replied:
So originally, I expected the array I am generating to end up being sparse, but I think I see now that this will not be the case. I'll remove the `filter`

JoostK commented on packages/compiler/src/shadow_css.ts:
```diff
@@ -650,3 +656,45 @@ function escapeBlocks(
+    const length = combined.length;
+    const contextSelector = contextSelectors.pop();
+    for (let i = 0; i < length; i++) {
+      combined[length * 2 + i] = combined[i] + ' ' + contextSelector;
+      combined[length + i] = contextSelector + ' ' + combined[i];
+      combined[i] = contextSelector + combined[i];
```
This loop is pretty tricky to wrap your head around, especially because of the last statement here that prepends `contextSelector` into `combined[i]`, whereas `combined[i]` has been used prior to it being prepended in the initial two statements. I think it would be more clear if `combined[i]` is first assigned into a variable.

petebacondarwin replied:
Will do - and I will add some more comments

JoostK commented on packages/compiler/src/shadow_css.ts:
```diff
@@ -650,3 +656,45 @@ function escapeBlocks(
+function combineHostContextSelectors(
+    hostMarker: string, contextSelectors: string[], otherSelectors: string): string {
+  const combined: string[] = [contextSelectors.pop() || ''];
+  while (contextSelectors.length) {
```
nit:

```suggestion
  while (contextSelectors.length > 0) {
```

JoostK commented on packages/compiler/src/shadow_css.ts:
```diff
@@ -260,15 +260,28 @@ export class ShadowCss {
+    return cssText.replace(_cssColonHostRe, (_, hostSelectors, rest) => {
+      if (hostSelectors) {
+        const parts = hostSelectors.split(',');
+        const r: string[] = [];
```
Can `r` be given a more descriptive name? Perhaps `rules` (I think)?

edit: I notice now this is moved code...

petebacondarwin replied:
I'll rewrite it as 

```ts
        const convertedSelectors =
            hostSelectors.split(',')
                .map(p => p.trim())
                .filter(p => !p)
                .map(p => this._colonHostPartReplacer(_polyfillHostNoCombinator, p, rest));
        return convertedSelectors.join(',');
```

petebacondarwin replied:
Oops! That doesn't work since `break` is different to `continue`...

petebacondarwin replied:
OK, I partially reverted my rewrite and changed the variable names to make them more descriptive.

JoostK reviewed (approved):
I am no subject expert on this matter but this looks correct to me.

AndrewKushnir reviewed (approved):
👍

AndrewKushnir commented:
[Presubmit](http://test/OCL:352898371:BASE:352898387:1611188089282:4e627a0).

AndrewKushnir commented:
@petebacondarwin presubmit indicated that there are some issues in IE11 on Win7, here is an error message:
```
TypeError: Syntax error in regular expression
```
The error is coming from the minified code, so there is no exact location available in logs :(

Please let me know if you would you be able to test it in IE11 or you want me to do more debugging and find the problematic RegExp?

petebacondarwin commented:
@AndrewKushnir - I suspect it must be that in IE11 the `RegExp` constructor cannot take a `RegExp` as its first argument: https://stackoverflow.com/a/40629349. I'll fix that.

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:352898371:BASE:353087761:1611263958776:c805d32c).

crisbeto commented on packages/compiler/src/shadow_css.ts:
```diff
@@ -260,15 +260,29 @@ export class ShadowCss {
+    return cssText.replace(_cssColonHostRe, (_, hostSelectors: string, otherSelectors: string) => {
+      if (hostSelectors) {
+        const convertedSelectors: string[] = [];
+        const hostSelectorArray = hostSelectors.split(',').map(p => p.trim());
```
This is a bit of an edge case, but the selector could have a comma inside quoted content, e.g. `[some-attr="foo,bar"]`. Do we have to account for cases like that?

petebacondarwin replied:
This code (for dealing with `:host`) is basically copied over with no change to its behaviour, so if it doesn't work now, it didn't work before. Generally, my understanding is that this emulation is not expected to be flawless. See https://github.com/angular/angular/blob/master/packages/compiler/src/shadow_css.ts#L24-L28

AndrewKushnir commented:
@petebacondarwin additional changes fixed issues in the presubmit and it's now passing. Thank you.

jessicajaniuk commented:
We had failures in Google3 with this PR and had to rollback. The string we found that was causing the failure was `-shadowcsscontext .full-snackbar`.

Caretaker note: We need a TGP once this fix is updated. The failing target is outside of our blueprint sample set.

petebacondarwin commented:
The use of `:host-context` without any selector was throwing an error. Specifically the app in G3  must have a style that looks like:

```css
:host-context .full-snackbar
```

or

```css
:host-context() .full-snackbar
```

To be honest, I am not sure what the correct behaviour is here. From my understanding of the [spec](https://drafts.csswg.org/css-scoping/#selectordef-host-context) it doesn’t make sense for this pseudo-class to have no selector.

If there is no selector then I would expect `:host-context()` to either match “any” ancestor, or perhaps even “no” ancestor. In the latter case the whole rule would match nothing. In the former the rule would be equivalent to just `.full-snackbar`.

I suspect that this is actually an error in the component’s styling (and perhaps we should report an error although this would be a breaking change) and they actually meant `:host-context(.full-snackbar)` or just `.full-snackbar`?

Before this PR, the compiler used to generate the equivalent of `:host .full-snackbar`. See https://stackblitz.com/edit/angular-ivy-xiqjs7?file=src%2Fapp%2Fapp.component.html.

So I have updated the code here to reproduce that same behaviour.

JoostK commented on packages/compiler/test/shadow_css_spec.ts:
```diff
@@ -248,6 +248,13 @@ import {normalizeCSS} from '@angular/platform-browser/testing/src/browser_util';
             ]);
       });
+
+      it('should handle :host-context with no ancestor selectors', () => {
```
Could you include a comment here with the context you provided in https://github.com/angular/angular/pull/40494#issuecomment-766096291, so that it's clear this behaviour is chosen for backwards compatibility, not spec compliancy per se?

petebacondarwin replied:
Done! Thanks for the review

petebacondarwin commented:
I believe I have fixed the problem: Time for another TGP since the failing target was outside of our blueprint sample set.

petebacondarwin committed:
fix(compiler): handle `:host-context` and `:host` in the same selector

In `ViewEncapsulation.Emulated` mode the compiler converts `:host` and
`:host-context` pseudo classes into new CSS selectors.

Previously, when there was both `:host-context` and `:host` classes in a
selector, the compiler was generating incorrect selectors. There are two
scenarios:

* Both classes are on the same element (i.e. not separated). E.g.
  `:host-context(.foo):host(.bar)`. This setup should only match the
  host element if it has both `foo` and `bar` classes. So the generated
  CSS selector should be: `.foo.bar<hostmarker>`.
* The `:host` class is on a descendant of the `:host-context`. E.g.
  `:host-context(.foo) :host(.bar)`. This setup should only match the
  `.foo` selector if it is a proper ancestor of the host (and not on the
  host itself). So the generated CSS selector should be:
  `.foo .bar<hostmarker>`.

This commit fixes the generation to handle these scenarios.

Fixes #14349

AndrewKushnir commented on packages/compiler/test/shadow_css_spec.ts:
```diff
@@ -208,6 +212,70 @@ import {normalizeCSS} from '@angular/platform-browser/testing/src/browser_util';
+        expect(s('.outer :host-context() .inner {}', 'contenta', 'a-host'))
+            .toEqual('.outer [a-host] .inner[contenta] {}');
+      });
+<<<<<<< HEAD
```
@petebacondarwin it looks like there is a merge conflict. Could you please have a look?

(Note: this should not affect TGP though since it's a spec file)

petebacondarwin replied:
Sorry, I forgot to save the file

AndrewKushnir commented:
Started a new [Ivy presubmit](http://test/OCL:353536449:BASE:353536444:1611523962007:9fac4495) and [TGP](https://test.corp.google.com/OCL:353536449:BASE:353536504:1611524029538:887f07d5), will keep this thread updated.

petebacondarwin commented:
I ended up adding an extra fix, since it seems like we would have to go through the larger presubmit for the second fix and we might as well batch them.

AndrewKushnir commented:
Thanks for the quick fix @petebacondarwin, I've cancelled the previous TGP run and [started a new TGP](https://test.corp.google.com/OCL:353536449:BASE:353537075:1611524463918:16b03943) with the most recent changes from this PR (also [Ivy Blueprint run](https://test.corp.google.com/ui#id=OCL:353536449:BASE:353537048:1611524418791:6e257788)).

petebacondarwin commented on packages/compiler/test/shadow_css_spec.ts:
```diff
@@ -208,6 +212,67 @@ import {normalizeCSS} from '@angular/platform-browser/testing/src/browser_util';
+      // It is not clear what the behaviour should be for a `:host-context` with no selectors.
+      // Arguably it is actually an error that should be reported.
+      it('should handle :host-context with no ancestor selectors', () => {
+        expect(s('.outer :host-context .inner {}', 'contenta', 'a-host'))
```
Also, arguably, the `.outer` should match only inside the shadow DOM and not outside, which makes a selector like this very dubious IMO.

AndrewKushnir commented:
Quick update after running TGP: there are some cases when multiple selectors are used inside `:host-context`, for ex.:
```
:host-context(.foo, .bar) { padding: 0px; }
```
That would require additional investigation on whether we should support it (it doesn't seem to be a part of the spec) or drop support for it (thus making this a breaking change).

petebacondarwin committed:
fix(compiler): support multiple selectors in `:host-context()`

The previous commits refactored the `ShadowCss` emulator to support
desirable use-cases of `:host-context()`, but it dropped support
for passing a comma separated list of selectors to the `:host-context()` .

This commit rectifies that omission, despite the use-case not being
valid according to the ShadowDOM spec, to ensure backward compatibility
with the previous implementation.

AndrewKushnir commented:
[Presubmit #3](http://test/OCL:353536449:BASE:353885256:1611682907430:e6c4b149).

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/OCL:353536449:BASE:354223200:1611803551175:44604b1c).

AndrewKushnir commented:
Quick update after global presubmit: there is one failing target (not a flake) with a slight visual difference in screen shot tests. I will perform further investigation and will keep this thread updated...

AndrewKushnir commented:
@petebacondarwin I've extracted a part of CSS that is now processed differently:

```
:host-context(.toolbar--top) :host {}
```

Current (in master branch):
```
.toolbar--top[a-host] [a-host],
.toolbar--top [a-host] [a-host] {}
```

After the changes in this PR:
```
.toolbar--top [a-host] {}
```

petebacondarwin commented:
Thanks for running the presubmit and providing this breaking case @AndrewKushnir. I think this really is a situation where this PR is correct and the old behaviour is wrong.  The way to interpret the style in question:

```css
:host-context(.toolbar--top) :host {}
```

Is that it will match only if there is an element with class `toolbar--top` that is on either a full ancestor of the host element or the host itself with a full descendant of this element being the host element. In other words:

```csss
.toolbar--top[a-host] [a-host]
.toolbar--top [a-host]
```

Since the former of these two can never match, if taking the spec accurately, (you cannot have a host that is descended from itself) then only the latter selector is relevant. So we end up with

```css
.toolbar--top [a-host] {}
```

If we consider the selectors previously generated:

```css
.toolbar--top[a-host] [a-host]
.toolbar--top [a-host] [a-host]
```

There are two scenarios... 

1. The component is self-referential in some way, in which case the following DOM structures would match

```
.toolbar--top
  -> <my-component>
    -> other nodes
      -> <my-component>
```

and

```
<my-component>.toolbar--top
  -> other nodes
    -> <my-component>
```

2. The component does not reference itself at all, in which case these selectors would never match anything.

Clearly the application developer did not mean the second scenario. So it is possible that they meant the first scenario, although this is not valid re the spec. But if they really wanted to do that then I would imagine the could do something like:

```css
:host-context(.toolbar--top,<component-selector>) :host {}
```

where `<component-selector> is a way to select the host element directly - perhaps by its element name.

---


So I think that in this case we should fix up the downstream application rather than make any change to this PR. WDYT?

petebacondarwin commented:
This collection of examples seems to show that the implementation of `:host-context(xxx) :host` type selectors in this PR is at odds with what Chrome does.

https://stackblitz.com/edit/angular-ivy-wamhub?file=src%2Fapp%2Fstyles.ts

So either Chrome is also buggy in this regard or my understanding of the spec is wrong.

petebacondarwin commented:
Also I note that the failure is a style that is no longer being assigned but the new selector is more relaxed than the old selector: The new selector `.toolbar--top [a-host]` in this PR would select everything that the previous selectors would: `.toolbar--top[a-host] [a-host], .toolbar--top [a-host] [a-host]`.
```

AndrewKushnir commented:
FYI, the code of an app was updated in Google's codebase (it turned out that the CSS rule was unused) and I've started a new [Global Presubmit](https://test.corp.google.com/OCL:357628614:BASE:357628644:1613440431175:7416dd9d) to verify that there are no similar cases and the change can be safely submitted. Will keep this thread updated.

AndrewKushnir commented:
@petebacondarwin FYI presubmits are successful for the changes in this PR, we can proceed with merging (plz add the "merge" label if it's ready to go)!

petebacondarwin commented:
That's great news @AndrewKushnir - thanks for helping to shepherd this through.
